ivy/functional/backends/jax/device.py

```diff
@@ -50,53 +57,69 @@ def dev(
 
 def to_device(
     x: JaxArray,
-    device: jaxlib.xla_extension.Device,
+    device: Union[jaxlib.xla_extension.Device, ivy.Device],
     /,
     *,
     stream: Optional[int] = None,
     out: Optional[JaxArray] = None,
 ):
-    if device is not None:
-        cur_dev = as_native_dev(dev(x))
-        if cur_dev != device:
-            x = jax.device_put(x, as_native_dev(device))
-    return x
+    # TODO: implement stream and out
+    return _to_device(x, device=device)
 
 
 # this is a non-wrapped function used to place JAX arrays on respective devices,
 # since if we use to_device, it will return ivy.array which is not desirable
 def _to_device(x, device=None):
     if device is not None:
-        cur_dev = as_native_dev(dev(x))
+        device = as_native_dev(device)
+        cur_dev = dev(x, as_native=True)
         if cur_dev != device:
-            x = jax.device_put(x, as_native_dev(device))
+            # Uses async dispatch, data commited only when required
+            x = jax.device_put(x, device)
     return x
 
 
 def as_ivy_dev(device, /):
+    # doesn't check if device is a valid device
     if isinstance(device, str):
         return ivy.Device(device)
-    if device is None:
-        return None
-    p, dev_id = (device.platform, device.id)
-    if p == "cpu":
-        return ivy.Device(p)
-    return ivy.Device(p + ":" + str(dev_id))
+    if is_native_dev(device):  # no duck typing
+        p, dev_id = (device.platform, device.id)
+        if p == "cpu" and dev_id == 0:
+            return ivy.Device(p)
+        return ivy.Device(p + ":" + str(dev_id))
+    else:
+        raise ivy.utils.exceptions.IvyException(
+            f"Cannot convert {device} to an ivy device. Expected a "
+            f"jaxlib.xla_extension.Device or str, got {type(device)}"
+        )
 
 
 def as_native_dev(device, /):
-    if not isinstance(device, str):
+    # checks if device is a valid device internally
+    if is_native_dev(device):
         return device
-    dev_split = ivy.Device(device).split(":")
-    device = dev_split[0]
-    if len(dev_split) > 1:
-        idx = int(dev_split[1])
+    if isinstance(device, str):
+        dev_split = ivy.Device(device).split(":")
+        device = dev_split[0]
+        if len(dev_split) > 1:
```

Comment:
can we also get the arbitrary devices thing working as well as discussed for this function and other functions in the backends?

Reply:
I think that is more of a device placement issue, not a converting one. We shouldn't be handling arbitrary device logic in `to_ivy_device` or `to_native_device` but rather during device placement (which makes me wonder if https://github.com/unifyai/ivy/pull/16101 is a good case for this). Maybe a function `is_real_device` or `to_nearest_real_device` is more suitable for such operations.

Reply:
Oh okay, makes sense! Thanks for clarifying, yeah we can handle this, not necessarily in the device shifting decorators but later. I'm not sure about the utility of `is_real_device` or `to_nearest_real_device`, so probably it could just be a private helper 👍